### ISSUE TITLE
Introduce "Treat all warnings as errors" policy for all project

### ISSUE THREAD

**Issue by andronov-alexey** [CoatiSoftware/Sourcetrail#814]
There's many errors during compilation and I'd like to remove them and enforce the policy in cmake

```
1>C:\Sourcetrail\src\src\lib_gui\qt\element\code\QtCodeArea.cpp(246,57): warning C4267: 'initializing': conversion from 'size_t' to 'int', possible loss of data
1>C:\Sourcetrail\src\src\lib_gui\qt\element\code\QtCodeArea.cpp(246,57): warning C4267: 'initializing': conversion from 'size_t' to 'const int', possible loss of data
1>C:\Sourcetrail\src\src\lib_gui\qt\element\code\QtCodeArea.cpp(278,44): warning C4267: 'return': conversion from 'size_t' to 'int', possible loss of data
1>C:\Sourcetrail\src\src\lib_gui\qt\element\code\QtCodeArea.cpp(430,88): warning C4267: 'argument': conversion from 'size_t' to 'int', possible loss of data
1>C:\Sourcetrail\src\src\lib_gui\qt\element\code\QtCodeArea.cpp(451,30): warning C4267: '=': conversion from 'size_t' to 'int', possible loss of data
```

Is everybody ok if I submit PR with this functionality?

**Comment by egraether** [comment]
yes

**Comment by mlangkabel** [comment]
Which compiler are those warnings based on? Will we integrate this into out CI? Otherwise we will run into issues with people that work with a different compiler. 

Furthermore, can you somehow exclude external code from this policy?

**Comment by andronov-alexey** [comment]
> Which compiler are those warnings based on?

First it would be `MSVC`, then probably `Clang`
> Will we integrate this into out CI

Yes, automatically. If during compilation a warning was generated - build will fail
> Furthermore, can you somehow exclude external code from this policy?

I guess so. We can disable/modify the policy in cmake for third-party code and also control the behavior directly in code via `#pragma warning`

**Comment by mlangkabel** [comment]
@andronov-alexey: Can I assign you to this issue?

**Comment by andronov-alexey** [comment]
Sure, I hope it's not urgent and I can do it leisurely :)

**Comment by mlangkabel** [comment]
Of course! However, I would suggest that everytime you managed to fix a few warnings, you submit a pull request so that you minimize the chance of ending up with a lot of merge conflicts ;)

**Comment by Piraty** [comment]
`-Werror` is a flag to use in development, not in production / distributed sources.

Distributions and other people will have a hard time building on various different build hosts and compiler variants, throwing all kinds of warnings.
Since users don't care about compiler warnings, they will have to disable it anyway. Since devs will be only ones benefiting from this, only devs should use it.

Guard this by a DEVELOPER_MODE variable or alike, but **don't enable this unconditionally!**

I see #850 introduces this for a specific windows compiler, but what do you gain? Do you want to stay in sync with all possible compiler variants out there? Good luck.
